cpp/src/arrow/compute/kernels/cast.cc

```diff
@@ -688,6 +680,35 @@ class ListCastKernel : public CastKernelBase {
   std::unique_ptr<UnaryKernel> child_caster_;
 };
 
+// ----------------------------------------------------------------------
+// Null to other things
+
+class FromNullCastKernel : public CastKernelBase {
+ public:
+  explicit FromNullCastKernel(std::shared_ptr<DataType> out_type)
+      : CastKernelBase(std::move(out_type)) {}
+
+  Status Call(FunctionContext* ctx, const Datum& input, Datum* out) override {
+    DCHECK_EQ(Datum::ARRAY, input.kind());
+
+    const ArrayData& in_data = *input.array();
+    DCHECK_EQ(Type::NA, in_data.type->id());
+    auto length = in_data.length;
+
+    // A ArrayData may be preallocated for the output (see InvokeUnaryArrayKernel),
+    // however, it doesn't have any actual data, so throw it away and start anew.
+    std::unique_ptr<ArrayBuilder> builder;
+    RETURN_NOT_OK(MakeBuilder(ctx->memory_pool(), out_type_, &builder));
+    RETURN_NOT_OK(builder->Reserve(length));
```

Comment:
Isn't this implied by `AppendNulls(lenght)`?

Reply:
Probably, yes. Will try to remove it.